rfcs/subsuites.md

```diff
@@ -0,0 +1,348 @@
+
+1. The tests are split into test groups, according to the
+   configuration.
+
+The `--test-groups` command line argument changes the behaviour here
+slightly; it directly provides a list of tests that will run, and
+names groups for the tests in the format `{group_name:
+[tests_ids]}`. This can be used in combination with the `--include`
+command line argument to subset the tests in the file, but trying to
+specify more tests than in the file will cause an error.
+
+The current metadata processing means that we need to read the
+metadata files once per subsuite (since we resolve the metadata
+properties when reading the file, and the metadata can be different
+for different subsuites).
+
+The simplest solution here is to read the `MANIFEST.json` files once,
+apply the top-level filters that apply to all subsuites (e.g. those
+from `--include`), and then run the remaining steps for each subsuite,
+adding the subsuite include filters after parsing the metadata. This
+isn't the optimal solution, since we can end up reading metadata
+multiple times even though it's not used. However further
+optimisations are possible (e.g. separating out the parse and
+evaluation phase of reading metadata, as we do with the update tool,
+making the top-level filters per-subsuite to avoid reading metadata
+files for tests won't be included in the subsuite). However these
+optimistations should be possible without backwards-incompatible
+changes.
+
+Since the `--test-groups` file is directly providing a list of tests
+that run, rather than a filter, the format of that file is modified in
+a backwards compatible way so that instead of the keys being just
+group name, they are of the form `<subsuite_name>:<group_name>`, with
+the part up to and including the colon omitted for the default subsuite.
+
+Running individual subsuites will be possible with the `--subsuite`
+command line argument. This can be provided multiple times to run
+multiple subsuites.
+
+#### wptreport.json
+
+The wptreport.json format needs to be updated to allow specifying
+which subsuite each test belongs to. To do this two principal changes
+are proposed:
+
+1. A new top-level key `subsuites` which contains a mapping from
+   subsuite name to the `run_info` data specific to that subsuite.
+
+1. Entries in the `results`, `asserts`, `lsan_leaks`, and `mozleak`
+   top-level fields get a `subsuite` property which is the name of a
+   subsuite.
+
+#### Metadata update
+
+The automatic metadata update works by collecting test results from
+multiple files, grouped by their `run_info` data. This can naturally
+be extended to support subsuites, just by extending the processing to
+create a `run_info` entry per subsuite rather than assuming the
+`run_info` is the same for all tests in a file.
+
+The main difficulty here is that the metadata expression language
+doesn't have a concept of `null`. So we need to represent the default
+subsuite name as the empty string in `run_info`, or add a `null`
```

Comment:
I think we would prefer to add a `null` concept. We already such issue for adding flag specific config in run_info. And we are using the empty string to represent non flag specific suite.